design-documents/scalars-for-templates.md

```diff
@@ -0,0 +1,177 @@
+E-mail templates - scalars received by calling certain model layer services.
+Thus templates will only receive data without the possibility to execute any code.
+
+#### What needs to be done
+###### Allow access to arrays' values
+We need to be able to use hashmaps in templates just as we use objects right now e.g.
+```
+{{var customer.name }}
+```
+_Magento\Framework\Filter\Template_ does not allow that right now so
+it's _getVariable()_ method must be updated to allow just that.
+ 
+To make these changes easier to adopt
+```
+{{var customer.getName() }}
+```
+ 
+Will be equal to accessing the 'customer' array by 'name' key.
+
+###### Remove access to $this
+Since $this use in E-mail templates refers to _Magento\Email\Model\Template_
+it has to be removed as a default variable passed to templates.
+ 
+Instead of using methods of _Magento\Email\Model\Template_ accessed with _$this_
+like
+```
+$this.getUrl()
+```
+introduce new directives.
+
+###### Do not accept objects as template variables
+_Magento\Framework\Filter\Template::setVariables()_ needs to filter out all objects and
+leaves only scalars and arrays of scalars.
+ 
+Also it's _getVariable()_ methods needs to updated to provide access only to scalar variables.
+ 
+To preserve backward compatibility instances of _DataObject_ will be accepted as well, then recursively their
+_getData()_ methods will be called to get a graph of scalars to use in templates. 
+
+###### Update core E-mail templates
+All E-mail templates present in core Magento must be updated to use values instead
+of calling methods.
+ 
+When providing data to templates we need not to forget about extensibility and
+pass redundant data - for instance if a 'new account' email uses only customer's
+full name pass the whole customer info retrieved from _Customer::getData()_.
+
+###### Make introducing new template directives easy for developers
+Right now one can only add a directive by extending the _Template_ class and introducing
+a new _\<directive name\>Directive()_ method - that is not a good way for developers
+to introduce new directives - what if we have 2 separate extensions that want to introduce
+2 different directives?
+ 
+To solve this next SPI will be introduced:
+```php
+DirectiveProcessorInterface
+{
+    /**
+     * Unique name of this directive.
+     *
+     * @return string
+     */
+    public function getName(): string;
+
+    /**
+     * Process values given to the directory and return rendered result.
+     *
+     * @param string|null Main parameter.
+     * @param string[] Additional parameters.
+     * @param string|null $html
+     * @return string
+     */
+    public function process(?string $value, ?string $html, array $parameters): string;
+    
+    /**
+     * Subdirectories that can be used within this directory's HTML.
+     * 
+     * @return DirectiveProcessorInterface[]|null
+     */
+    public function getSubDirectives(): ?array;
```

Comment:
Do we need to limit directives inside a directive? Can we just process recursively? Could you provide some examples, please?

Reply:
'else' directive would only be valid inside an 'if' directive